File: ZOJ/1549.cpp
/*
#Problem ID: 1549
#Submit Time: 2013-07-02 09:01:23
#Run Time: 10
#Run Memory: 188
#ZOJ User: calvinxiao
*/

/**
Calvin Xiao
calvin.xiao at scaurugby.com
*/
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <stack>
#include <queue>
#include <set>
#include <map>
#include <algorithm>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <cstdio>

using namespace std ;

#define REP(i, n) for (int i = 0; i < (int)(n); ++i)
#define REPN(i, a, n) for (int i = (int)(a); i <= (int)(n); ++i)
#define REPI(itr, c) for(__typeof((c).begin()) itr = (c).begin(); itr != (c).end(); itr++)

long long func(long long n, long long num, vector<long long> &ln) {
	int nLength = ln.size();
	long long ret = 0, ten = 10;
	while (num * ten <= n) {
		long long nt = num * ten;
		long long L = 0, R = ten - 1;
		long long toadd = -1;
		if (nt + ten - 1 <= n)
            toadd = ten - 1;
        else
            while (L <= R) {
                long long mid = L + R >> 1;
                if (num * ten + mid <= n) {
                    toadd = mid;
                    L = mid + 1;
                } else
                    R = mid - 1;
            }
		ret += toadd + 1;
		ten *= 10LL;
	}
	return ret + 1;
}

long long get(long long n, long long k, vector<long long> &ln, vector<long long> &lk) {
	long long ret = 0;
	int kLength = lk.size();
	long long num = 0;
	for (int i = 0; i < kLength; ++i) {
		num *= 10;
		int start = 0;
		if (i == 0)
			start = 1;
		for (long long j = start; j < lk[i]; ++j) {
			ret += func(n, num+j, ln);
		}
		num += lk[i];
	}
	return ret + kLength;
}


int main()
{
    //freopen("amusing.in", "r", stdin);
    //freopen("amusing.out", "w", stdout);
    long long K, M;
    while (~scanf("%lld%lld", &K, &M)) {
    	vector<long long> lk;
    	long long kk = K;
    	while (kk != 0)
    		lk.push_back(kk % 10LL), kk /= 10LL;
    	reverse(lk.begin(), lk.end());
    	long long L = K, R = (long long)1e18;
    	long long ans = 0;
    	while (L <= R) {
    		long long mid = L + R >> 1;
    		vector<long long> ln;
    		long long nn = mid;
    		while (nn != 0)
    			ln.push_back(nn % 10LL), nn /= 10LL;
    		long long ret = get(mid, K, ln, lk);
    		if (ret >= M) {
    			R = mid - 1;
    			if (ret == M)
    				ans = mid;
    		} else
    			L = mid + 1;
    	}
    	printf("%lld\n", ans);
    }
    return 0;
}

File: ZOJ/1549.py
#Problem ID: 1549
#Submit Time: 2013-07-02 08:59:07
#Run Time: 810
#Run Memory: 320
#ZOJ User: calvinxiao

import sys

#sys.stdin = open("amusing.in", "r")
#sys.stdout = open("amusing.out", "w")

def func(n, num, ln):
    nLength = len(ln)
    ret = 0
    ten = 10
    while num * ten <= n:
        #print num, ten, n
        nt = num * ten
        L, R = 0, ten-1
        toadd = -1
        if nt + ten - 1 <= n:
            toadd = ten - 1
        else:
            while L <= R:
                mid = L + R >> 1
                if nt + mid <= n:
                    toadd = mid
                    L = mid + 1
                else:
                    R = mid - 1
        ret += toadd + 1
        ten *= 10
    return ret + 1

def get(n, k, ln, lk):
    ret = 0
    kLength = len(lk)
    num = 0
    for i in xrange(kLength):
        num *= 10
        start = 0
        if i == 0:
            start = 1
        for j in xrange(start, lk[i]):
            ret += func(n, num+j, ln)
        num += lk[i]
        #print num
    return ret + kLength
            
#print get(100000000888888879, 100000001)

while 1:
    line = sys.stdin.readline()
    if not line:
        break
    K, M = map(int, line.split())
    lk = map(int, list(str(K)))
    L, R = K, int(1e18)
    ans = 0
    while L <= R:
        mid = (L + R) >> 1
        ln = map(int, list(str(mid)))
        ret = get(mid, K, ln, lk)
        #print L, R, mid, ret
        if ret >= M:
            R = mid - 1
            if ret == M:
                ans = mid
        else:
            L = mid + 1

    print ans
